hashicorp/terraform
# Implement migration away from TFC; unify migration code paths a bit more

Previously, we didn't support migrating away from the new TFC integration to an old-school state backend via a normal `terraform init -migrate-state`. This PR adds that capability.

Well, really all it _adds_ is some alternate message text for various stages of the process; most of the heavy lifting happens by _removing_ the TFC-specific reimplementation of the `backendMigrateState` wrapper function. 

When migrating onto or off of TFC, we need to take some slightly different signals into account to determine which path we're taking. But the ultimate decisions about which path to take end up being almost exactly the same (with the notable exception of multi-to-multi-with-renames); it's just that we need to check different values. 

So instead of running through the logic twice, this PR adds a couple of conditionals to the beginning of `backendMigrateState` to translate the TFC-specific signals about what's going on into the format that the rest of the function expects. Once that's done, the normal four-way switch statement works almost unmodified. 

## How to test this

The newly supported behaviors all involve:

1. Setting up + initting the cloud block, then performing a run to create some state.
2. Removing the cloud block, replacing it with either a `backend` block or nothing (implicit local backend), and re-running init with `-migrate-state`. 

In all cases, `terraform workspace list` should show you at least the workspaces you intended to migrate (sometimes with the addition of `default`), and you should be able to see the state data if you dig into the appropriate storage. The state should also remain in the TFC workspaces unmodified, since migration only copies; the user is in charge of deletion. 

Some cases worth looking at: 

- `cloud` with `name`, migrating to (nothing, consul, etc.): any existing workspaces in the new backend still exist (including `default`, even if it's empty), and it should add a named workspace with the original name. The new workspace is selected. 
    - If a workspace with the target name already existed in the new backend, it should ask you which state to use. 
- `cloud` with `tags` and multiple workspaces migrating to (nothing, consul, etc.): existing workspaces in backend still exist, new workspaces are added for each workspace from TFC. 
    - For any workspace names that already exist, it should ask which state to use. 
- `cloud` with `tags` and only one matching workspace migrating to (nothing, consul, etc.): Should behave like migrating with `name`.
- `cloud` with `name` migrating to a backend that only supports one state at a time: Should migrate the TFC state to the `default` workspace in the new backend.
- `cloud` with `tags` and multiple workspaces migrating to a single-state backend: Should confirm that you're okay with migrating only the current selected workspace to `default`, and then do it. 
- `cloud` with `tags` and only one workspace matching, migrating to single-state: Should behave like `name` -> single-state. (Doesn't need to confirm that only the current workspace will ride along, since there's only one.)
